2 Niagara Fundamentals Exercise - Multiplexer/multiplexer-rt/src/com/tridiumuniversity/multiplexer/BMultiplexer.java

```diff
@@ -0,0 +1,284 @@
+        for (int i = inputSteps.length - 1; i>-1; i--)
+        //
+        {
+            if (inInputCount >= inputSteps[i])
+            {
+                effInputStep = i;
+                log.fine("Effective input step: " + effInputStep);
+                break;
+            }
+        }
+    }
+
+
+    /* Calculates the switchSum based on the combination of switches currently set to "true" */
+    private void calcSwitchSum()
+    {
+        switchSum = 0; // reset the switchSum
+
+        /* process all the "in" inputs */
+        for (int i = 0; i<sInputs.size(); i++)
+        {
+            if (((BBoolean) sInputs.get(i).getDefaultValue()).getBoolean()) // get input as boolean. If true, add a numeric value to the switchSum
+            {
+                String propertyName = sInputs.get(i).getName(); // get the property's name
+                int sNumber = Integer.parseInt(propertyName.replace("s", "")); // capture the number from the property's name
+
+                int switchValue = switchValues[sNumber]; // figure out the switch's numeric value based on the switch number
+
+                if (sNumber <= effInputStep - 1) // if the switch number is an "active" switch...
+                {
+                    switchSum += switchValue; // apply the numeric value towards the sum
+                }
+            }
+        }
+
+        setUsableSwitches(effInputStep);
+        setSwitchTotal(switchSum);
+    }
+
+
+    /* updates the out property */
+    private void updateOut()
+    {
+        log.info("Elevated Property: " + "in" + switchSum); // FOR SOME REASON THIS LOGS IN THE WB CONSOLE WHEN SWITCHING VIEWS
+
+        setOut(((BBoolean) getProperty("in" + switchSum).getDefaultValue()).getBoolean()); // e.g. a switchSum of 18 will pass in18 to "out"
+    }
+
+
+
+    private int inInputCount; // the number of "in" inputs in the multiplexer
+    private int sInputCount; // the number of "s" inputs in the multiplexer
+
+    private ArrayList<Property> inInputs = new ArrayList<Property>(); // the "in" inputs as Property objects
+    private ArrayList<Property> sInputs = new ArrayList<Property>(); // the "s" inputs as Property objects
+
+    private final int[] inputSteps = {0, 2, 4, 8, 16, 32, 64, 128}; // values upon which more "s" inputs are needed/used
+    private int effInputStep; // the current amount of "s" inputs needed/used
+
+    private final int[] switchValues = {1, 2, 4, 8, 16, 32, 64};
+    private int switchSum; // switchSum is the value calculated by the combination of switches that are "true"
+    /* when a switch is "false", that switch counts a 0 towards the switchSum
+    s0 counts as 1 when "true", s1 counts as 2 when "true", s2 counts as 4 when "true", s3 counts as 8 when "true", etc
+    They work like a set of address DIP switches */
+
+    //private String elevatedInput; // the current inInput being sent to "out"
+
+    private static final Logger log = Logger.getLogger("multiplexer.multiplexer");
```

Comment:
Nice job on the logging! I'm sure that made debugging easier.

Reply:
It did!